### PR TITLE
Account for root surface transformation on the surfaces managed by the external view embedder.

### PR DESCRIPTION
**Pull request by chinmaygarde**
The earlier design speculated that embedders could affect the same
transformations on the layers post engine compositor presentation but before
final composition.

However, the linked issue points out that this design is not suitable for use
with hardware overlay planes. When rendering to the same, to affect the
transformation before composition, embedders would have to render to an
off-screen render target and then apply the transformation before presentation.
This patch negates the need for that off-screen render pass.

To be clear, the previous architecture is still fully viable. Embedders still
have full control over layer transformations before composition. This is an
optimization for the hardware overlay planes use-case.

Fixes b/139758641

### CODE REVIEW

**Comment by chinmaygarde** [comment]
cc @dnicoara

**Comment by dnicoara** [comment]
A couple of observations:
1) It looks like the FlutterCompositor APIs are now using the display coordinate system, rather than the surface coordinate system. Could this be documented in the API?

2) It looks like I also had to flip the image over x-axis. I'm not sure what caused this behavior. Any clues?

**Comment by chinmaygarde** [comment]
> It looks like the FlutterCompositor APIs are now using the display coordinate system, rather than the surface coordinate system. Could this be documented in the API?

Sure. I will document this. But does this match your expectations? I wasn't sure if the least unexpected behavior for embedders would be for them to be in control of applying this transformation themselves.

> It looks like I also had to flip the image over x-axis. I'm not sure what caused this behavior. Any clues?

The transformation is applied to the offset and the bounds of the layer. It is still embedder responsibility to render the transformed contents about the new offset. In fact, this is why the test is currently failing. My test compositor doesn't yet account for the transformation for the contents rendered by the embedder.

**Comment by chinmaygarde** [comment]
> It looks like I also had to flip the image over x-axis. I'm not sure what caused this behavior. Any clues?

Ok, I am convinced, this is bogus behavior in this patch. I am fixing it so you wont have to apply the offset about the x-axis. You will still have to apply to the transformation to the contents though.

**Comment by chinmaygarde** [comment]
Fixed the math. You no longer have to apply the flip over the x-axis unnecessarily. Also documented the behavior that setting the root surface transformation will affect layer offsets and sizes in the present callback. PTAL.

File: shell/platform/embedder/embedder_external_view_embedder.cc [hunk 1]
```diff
@@ -21,6 +21,19 @@ EmbedderExternalViewEmbedder::EmbedderExternalViewEmbedder(
...
+void EmbedderExternalViewEmbedder::SetSurfaceTransformationCallback(
+    SurfaceTransformationCallback surface_transformation_callback) {
+  surface_transformation_callback_ = surface_transformation_callback;
+}
+
+SkMatrix EmbedderExternalViewEmbedder::GetSurfaceTransformation() const {
+  if (!surface_transformation_callback_) {
+    return SkMatrix{};
+  }
+
```
[the remark this answers is not in the document]
Comment (chinmaygarde) [reply shell/platform/embedder/embedder_external_view_embedder.cc]: Disagree. We use the surface transformation to not only affect the size, but also the offset (of the root surface too). We will lose this information if we return identity here.

File: shell/platform/embedder/embedder_external_view_embedder.cc [hunk 2]
```diff
@@ -43,11 +56,31 @@ static FlutterBackingStoreConfig MakeBackingStoreConfig(const SkISize& size) {
...
+static SkISize TransformedSurfaceSize(const SkISize& size,
+                                      const SkMatrix& transformation) {
+  const auto source_rect = SkRect::MakeWH(size.width(), size.height());
+  auto transformed_rect = SkRect::MakeEmpty();
+
+  if (!transformation.mapRect(&transformed_rect, source_rect)) {
+    // TODO(chinmaygarde): This has the opportunity to become super spammy. Move
+    // this to the begin frame call so that embedders get just one warning per
+    // frame.
+    FML_LOG(ERROR) << "Embedder supplied surface transformation does not map a "
```
[the remark this answers is not in the document]
Comment (chinmaygarde) [reply shell/platform/embedder/embedder_external_view_embedder.cc]: On second thought, I think this block is incorrect. We can still return the bounding box of the transform box even when rect-ness is not maintained. Embedders can still apply transformations on their own.

[the remark this answers is not in the document]
Comment (chinmaygarde) [reply shell/platform/embedder/embedder_external_view_embedder.cc]: Fixed.

File: shell/platform/embedder/embedder_external_view_embedder.cc [hunk 3]
```diff
@@ -139,18 +190,27 @@ static FlutterPlatformView MakePlatformView(
...
+                              const SkMatrix& surface_transformation) {
   FlutterLayer layer = {};
 
   layer.struct_size = sizeof(layer);
   layer.type = kFlutterLayerContentTypePlatformView;
   layer.platform_view = &platform_view;
 
-  layer.offset.x = params.offsetPixels.x();
-  layer.offset.y = params.offsetPixels.y();
+  const auto layer_bounds = SkRect::MakeXYWH(params.offsetPixels.x(),    //
```
[the remark this answers is not in the document]
Comment (chinmaygarde) [reply shell/platform/embedder/embedder_external_view_embedder.cc]: I am actually using those to format the args in a column :/

File: shell/platform/embedder/tests/embedder_a11y_unittests.cc [hunk 4]
```diff
@@ -15,6 +15,7 @@
 #include "flutter/shell/platform/embedder/embedder.h"
 #include "flutter/shell/platform/embedder/tests/embedder_config_builder.h"
 #include "flutter/testing/testing.h"
+#include "third_party/skia/include/core/SkMatrix.h"
```
[the remark this answers is not in the document]
Comment (chinmaygarde) [reply shell/platform/embedder/tests/embedder_a11y_unittests.cc]: Not necessary. Removed.

**Comment by chinmaygarde** [comment]
More unit-tests have been added because @dnicoara was mentioning that root surface transformations may not be consistent between the compositor and non-compositor use cases. They were consistent but now we have unit-tests to verify this.

**Comment by chinmaygarde** [comment]
This is good to go. The coordinate space for all OpenGL backing stores is bottom left origin while that for Flutter uses top-left origin. This is consistent with the previous behavior of treating the root OpenGL surface as having a bottom left origin.

**Comment by dnicoara** [comment]
Thanks, it did work as expected for the 270 degree clockwise rotation. Would it be possible to add tests for other rotations (90, 180)? For example, when I used a 90 degree rotation I didn't get the content upside-down, it was still right side-up (though it's possible I messed up and forgot to update something).

**Comment by chinmaygarde** [comment]
Fixes https://github.com/flutter/flutter/issues/39339

**Pull request merged**